### PR TITLE
Docs/grpc plugins

### PR DESCRIPTION
**Pull request by tlblessing**
https://konghq.atlassian.net/browse/DOCS-1158

This PR adds a deep dive comparison of the Kong gRPC plugins written by Aaron Miller and other dev SMEs. The topic is in the Plugins TOC in the KE TOC. There isn't a Plugins section in the CE docs (non-hub) TOC. Added See also links to plugin hub docs to new intro topic that compares the gRPC plugins and walks through their set up and testing.

This PR does not address the specific issues mentioned for the gateway plugin in https://konghq.atlassian.net/browse/FTI-1750. That should be addressed in a subsequent PR.

Old ticket also addressed:

https://konghq.atlassian.net/browse/DOCS-625

Note that there doesn't seem to be a logical order for the nav plugin TOC, most of which links to plugin hub docs. Need to discuss with PM as to plugin docs overhaul and centralization.

Direct review links:

https://deploy-preview-2215--kongdocs.netlify.app/enterprise/2.1.x/plugins/grpc/

plugin hub docs:

https://deploy-preview-2215--kongdocs.netlify.app/hub/kong-inc/grpc-gateway/#see-also
https://deploy-preview-2215--kongdocs.netlify.app/hub/kong-inc/grpc-web/#see-also

### CODE REVIEW

**Comment by lena-larionova** [comment]
Should also update the supported versions in grpc-Web - it currently doesn't list Kong Enterprise. A quick test with Kong EE 2.1 beta tells me they both exist in that version, and looking over the FTI [changelog](https://konghq.atlassian.net/wiki/spaces/FTT/pages/8356124/Kong+Enterprise+-+Changelog+Features+Release+Schedule#KongEnterprise-Changelog,Features&ReleaseSchedule-KongEnterprise1.3.0.1), grpc support was adding to Kong EE in 1.3.

**Comment by tlblessing** [comment]
> Should also update the supported versions in grpc-Web - it currently doesn't list Kong Enterprise. A quick test with Kong EE 2.1 beta tells me they both exist in that version, and looking over the FTI [changelog](https://konghq.atlassian.net/wiki/spaces/FTT/pages/8356124/Kong+Enterprise+-+Changelog+Features+Release+Schedule#KongEnterprise-Changelog,Features&ReleaseSchedule-KongEnterprise1.3.0.1), grpc support was added to Kong EE in 1.3.

Done in commit e56f017

**Comment by tlblessing** [comment]
@aaronhmiller  and @lena-larionova please review when you get a chance. Thank you!

File: app/_hub/kong-inc/grpc-web/index.md [hunk 1]
```diff
@@ -45,13 +49,18 @@ params:
...
 ## Purpose
 
-A service that presents a gRPC API can be used by clients written in many languages, but the network specifications are oriented primarily to connections within a datacenter. [gRPC-Web] lets you expose this API to the Internet so that it can be consumed by browser-based JS apps.
+A service that presents a gRPC API can be used by clients written in many languages,
+but the network specifications are oriented primarily to connections within a datacenter.
+[gRPC-Web] lets you expose this API to the Internet so that it can be consumed by browser-based JS apps.
 
-This plugin translates requests and responses between [gRPC-Web] and ["real" gRPC](https://github.com/grpc/grpc).  Supports both HTTP/1.1 and HTTP/2, over plaintext (HTTP) and TLS (HTTPS) connections.
+This plugin translates requests and responses between [gRPC-Web] and
+["real" gRPC](https://github.com/grpc/grpc). The plugin supports both HTTP/1.1
```
Comment (lena-larionova) [inline app/_hub/kong-inc/grpc-web/index.md:57]:
```suggestion
[gRPC](https://github.com/grpc/grpc). The plugin supports both HTTP/1.1
```
"real" isn't needed here, it's implied in the name.

File: app/_hub/kong-inc/grpc-web/index.md [hunk 2]
```diff
@@ -45,13 +49,18 @@ params:
 
 ## Purpose
 
-A service that presents a gRPC API can be used by clients written in many languages, but the network specifications are oriented primarily to connections within a datacenter. [gRPC-Web] lets you expose this API to the Internet so that it can be consumed by browser-based JS apps.
+A service that presents a gRPC API can be used by clients written in many languages,
+but the network specifications are oriented primarily to connections within a datacenter.
```
Comment (lena-larionova) [inline app/_hub/kong-inc/grpc-web/index.md:53]: I'm not sure how to rephrase this, but it's hard to read as-is. The basic gist of this phrase is: you have a service with a gRPC API, you want to expose it to the internet, but the standard network specs don't provide that flexibility. 

Can we turn it into something more readable?

File: app/_hub/kong-inc/grpc-web/index.md [hunk 3]
```diff
@@ -45,13 +49,18 @@ params:
...
 
-This plugin translates requests and responses between [gRPC-Web] and ["real" gRPC](https://github.com/grpc/grpc).  Supports both HTTP/1.1 and HTTP/2, over plaintext (HTTP) and TLS (HTTPS) connections.
+This plugin translates requests and responses between [gRPC-Web] and
+["real" gRPC](https://github.com/grpc/grpc). The plugin supports both HTTP/1.1
+and HTTP/2, over plaintext (HTTP) and TLS (HTTPS) connections.
 
 ## Usage
 
-This plugin should be enabled on a Kong `Route` that serves the `http(s)` protocol but proxies to a `Service` with the `grpc(s)` protocol.
+This plugin should be enabled on a Kong `Route` that serves the `http(s)` protocol
```
Comment (lena-larionova) [inline app/_hub/kong-inc/grpc-web/index.md:62]:
```suggestion
This plugin should be enabled on a Kong Route that serves the `http(s)` protocol
```

File: app/_hub/kong-inc/grpc-web/index.md [hunk 4]
```diff
@@ -45,13 +49,18 @@ params:
...
+but proxies to a `Service` with the `grpc(s)` protocol.
```
Comment (lena-larionova) [inline app/_hub/kong-inc/grpc-web/index.md:63]:
```suggestion
but proxies to a Service with the `grpc(s)` protocol.
```

File: app/_hub/kong-inc/grpc-web/index.md [hunk 5]
```diff
@@ -91,9 +100,16 @@ $ curl -XPOST localhost:8001/routes/web-service/plugins \
...
+This minimal setup works for the default varieties of the [gRPC-Web protocol],
+which use ProtocolBuffer messages either directly in binary or with base64-encoding.
+The related `Content-Type` headers are `application/grpc-web` or `application/grpc-web+proto`
+for binary, and `application/grpc-web-text` or `application/grpc-web-text+proto` for text.
 
-If we wish to use JSON encoding, we have to provide the gRPC specification in a .proto file, which needs to be installed in the Kong node running the plugin.  A path starting with a `/` is considered absolute, otherwise it will be interpreted relative to the Kong's prefix (`/usr/local/kong/` by default).  For example:
+If you want to use JSON encoding, you have to provide the gRPC specification in
+a `.proto` file, which needs to be installed in the Kong node running the plugin.
+A path starting with a `/` is considered absolute; otherwise, it will be interpreted
+relative to the Kong's prefix (`/usr/local/kong/` by default). For example:
```
Comment (lena-larionova) [inline app/_hub/kong-inc/grpc-web/index.md:112]:
```suggestion
relative to the Kong node's prefix (`/usr/local/kong/` by default). For example:
```

File: app/enterprise/2.1.x/plugins/grpc.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
+---
+title: Introduction to Kong gRPC Plugins
+---
+
+Before we dive into the specifics of how to configure Kong's gRPC plugins, let's
+first discuss the advantages of the protocol. Unlike JSON, [gRPC](https://en.wikipedia.org/wiki/GRPC)
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/plugins/grpc.md:6]: This first sentence sounds more like a blog post than doc. Can you rephrase to make it sound less casual?

File: app/enterprise/2.1.x/plugins/grpc.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+is a binary protocol, using [protobuf](https://en.wikipedia.org/wiki/Protocol_Buffers)
+definitions to instruct how the data is marshalled and unmarshalled. Because
+binary data is used instead of text, it's a more efficient way to transmit data
+over a network. But this also makes gRPC harder to work with because inspecting
+what went wrong is more challenging. Additionally, few clients natively handle gRPC.
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/plugins/grpc.md:11]:
```suggestion
over a network. However, this also makes gRPC harder to work with, because inspecting
what went wrong is more challenging. Additionally, few clients natively handle gRPC.
```

File: app/enterprise/2.1.x/plugins/grpc.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+- [gRPC-Gateway](/hub/kong-inc/grpc-gateway)
+- [gRPC-Web](/hub/kong-inc/grpc-web)
+
+The gRPC-Gateway plugin allows you to send JSON requests to a gRPC service. A
+specially configured `.proto` file handles the conversion of the JSON request
+into one that the gRPC service can handle. This allows you to expose RESTful-style
+interfaces that talk to a gRPC service.
+
+The gRPC-Web plugin allows you to interact with a gRPC service from a browser.
+Instead of presenting a RESTful-type call, however, you POST data to the same
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/plugins/grpc.md:23]:
```suggestion
Instead of presenting a RESTful-type call, you POST data to the same
```
Not sure the "however" really belongs here, up to you to rephrase otherwise if needed.

File: app/enterprise/2.1.x/plugins/grpc.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+gRPC service endpoint that the protobuf defines.
+
+For flexibility and compatibility with RESTful expectations, the gRPC-Gateway
+plugin offers more configurability, whereas the gRPC-Web plugin adheres more
+directly to the protobuf specification.
+
+Let's walk through setting up each plugin so you can see how they work.
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/plugins/grpc.md:30]: This line isn't necessary.
```suggestion
```

Comment (tlblessing) [reply app/enterprise/2.1.x/plugins/grpc.md:30]: I think it's a decent lead-in that tells you you are going to try out both plugins.

File: app/_hub/kong-inc/grpc-web/index.md (hunk 2 above)
Comment (tlblessing) [reply app/_hub/kong-inc/grpc-web/index.md:53]: All I did was wrap the original text. The scope of this PR is the new non-hub write-up, primarily. This sentence in conjunction with the next sentence makes sense to me.

Comment (tlblessing) [reply app/_hub/kong-inc/grpc-web/index.md:53]: I'm going to edit the second sentence a bit...

File: app/_hub/kong-inc/grpc-web/index.md [hunk 10]
```diff
@@ -157,3 +182,7 @@ As an extension, this plugin also allows "naked" JSON requests with POST method
 [lua-protobuf]: https://github.com/starwing/lua-protobuf
 [lua-cjson]: https://github.com/openresty/lua-cjson
 [lua-pack]: https://github.com/Kong/lua-pack
+
+## See also
```
[the remark this answers is not in the document]
Comment (tlblessing) [reply app/_hub/kong-inc/grpc-web/index.md:186]: Confluent standard is "See also" so I was just using that. https://docs.confluent.io/current/cloud/cli/command-reference/ccloud_logout.html#see-also

@GayleNeumann what do you think?

File: app/enterprise/2.1.x/plugins/grpc.md (hunk 6 above)
Comment (tlblessing) [reply app/enterprise/2.1.x/plugins/grpc.md:6]: Yeah it is bloggy. Will rework.

File: app/_hub/kong-inc/grpc-web/index.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (tlblessing) [reply app/_hub/kong-inc/grpc-web/index.md:186]: Also, there are a bunch of links above with no heading at all. I'm not sure why all of those are present. I can move them below whatever heading we decide on.

[the remark this answers is not in the document]
Comment (tlblessing) [reply app/_hub/kong-inc/grpc-web/index.md:187]: Discussed today in docs meeting. @GayleNeumann is fine with See also and Related information as appropriate.
